patches/runtime/0026-Pin-System.Security.Cryptography.Cng-to-single-versi.patch

```diff
@@ -0,0 +1,30 @@
+From c1322c37dac74e51e0bef7e35edf2db05fa97869 Mon Sep 17 00:00:00 2001
+From: dseefeld <dseefeld@microsoft.com>
+Date: Fri, 30 Oct 2020 16:25:21 +0000
+Subject: [PATCH] Pin System.Security.Cryptography.Cng to single version
+
+If System.Security.Cryptography.Cng is restored using the PVP
+version (5.0.0), this causes conflicts later in the restore which
+results in an error:
+   An item with the same key has already been added.
+   Key: (System.Security.Cryptography.Cng, 5.0.0)
+---
```

Comment:
@dseefeld @dagood does this mean something doesn't allow us to consume `System.Security.Cryptography.Cng` `5.0.0` to produce `System.Security.Cryptography.Cng` `5.0.0`?

Reply:
Yes, that's what I was seeing.  Here's the error I was seeing without this patch:
```
3000 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error : An item with the same key has already been added. Key: (System.Security.Cryptography.Cng, 5.0.0) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3001 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3002 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at System.Linq.Enumerable.ToDictionary[TSource,TKey](List`1 source, Func`2 keySelector, IEqualityComparer`1 comparer) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3003 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.LockFileBuilder.CreateLockFile(LockFile previousLockFile, PackageSpec project, IEnumerable`1 targetGraphs, IReadOnlyList`1 localRepositories, RemoteWalkContext context) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3004 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreCommand.ExecuteAsync(CancellationToken token) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3005 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.ExecuteAsync(RestoreSummaryRequest summaryRequest, CancellationToken token) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3006 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3007 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.CompleteTaskAsync(List`1 restoreTasks) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3008 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.RunAsync(IEnumerable`1 restoreRequests, RestoreArgs restoreContext, CancellationToken token) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3009 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.RunAsync(RestoreArgs restoreContext, CancellationToken token) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3010 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Build.Tasks.BuildTasksUtility.RestoreAsync(DependencyGraphSpec dependencyGraphSpec, Boolean interactive, Boolean recursive, Boolean noCache, Boolean ignoreFailedSources, Boolean disableParallel, Boolean force, Boolean forceEvaluate, Boolean hideWarningsAndErrors, Boolean restorePC, Boolean cleanupAssetsForUnsupportedProjects, ILogger log, CancellationToken canc     ellationToken) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]
3011 /src/tb-5.0GA/.dotnet/sdk/5.0.100-rc.1.20452.10/NuGet.RestoreEx.targets(10,5): error :    at NuGet.Build.Tasks.Console.MSBuildStaticGraphRestore.RestoreAsync(String entryProjectFilePath, IDictionary`2 globalProperties, IReadOnlyDictionary`2 options) [/src/tb-5.0GA/src/runtime.cf258a14b70ad9069470a108f13765e0e5988f51/Build.proj]

```

Reply:
Doesn't this happen for a lot of packages since we're consuming 5.0.0 prebuilt packages to build 5.0.0 source-built packages? Or is issue somehow specific to the Cng package?

Reply:
I don't know why this issue popped up for this specific package.  We could certainly try to remove the patch and see if the issue repros.  It may have been a "point in time" issue.

Reply:
Here's some more info from our old chat about it. dotnet/runtime seems to be doing something specific that aggravates some NuGet bug. It won't be a point in time issue, I doubt the NuGet issue will be fixed in our timeline. I was able to put together a simplified repro, but we dropped it because we probably don't actually need a fix here. (And we always want to cut things/work because time is not something we have much of.)

> dagood: [in response to a binlog] Yeah, good old RestoreTaskEx... another issue we have with static restore is https://github.com/dotnet/source-build/issues/1825 FWIW, but unrelated.
> 
> ...
>
> Seems like we might be running into https://github.com/NuGet/Home/issues/6795 ? But not sure why we would but the Microsoft build wouldn't...

> [10/30 9:51 AM] Davis Goodin
>     Hmm, noticed a few things earlier in those RestoreTaskEx logs:
> 
> ```
> Could not parse version  for LatestPackage System.Security.Cryptography.Cng, will use latest stable.
> ...
> Installing System.Security.Cryptography.Cng 5.0.0.
> ```
> 
> (And actually the lines do seem scrambled around that installing line, so not sure which project it's coming from.)

> ​[10/30 9:52 AM] Davis Goodin
>     I guess some other project is referring to the Cng package. Not totally sure that's important though.

> ​[10/30 9:55 AM] Davis Goodin
>     Hmm. Wonder if it's choking on this, where in the prod build we don't have a PVP value but in the tarball build we're overwriting it?
> https://github.com/dotnet/runtime/blob/7ef6d50b312217d2f7c17b9697891fa8ab98a19d/src/libraries/System.Security.Cryptography.Pkcs/src/System.Security.Cryptography.Pkcs.csproj#L639-L642
> 
> ```
>  <ItemGroup Condition="$(TargetFramework.StartsWith('netstandard2.0'))">
>  <PackageReference Include="System.Buffers" Version="$(SystemBuffersVersion)" />
>  <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
>  <PackageReference Include="System.Security.Cryptography.Cng" Version="$(SystemSecurityCryptographyCngVersion)" />
> ```
> 

> ​[10/30 9:57 AM] Davis Goodin
>     For a sanity check... does the PVP being passed in in your local build have SystemSecurityCryptographyCngVersion defined?

> [10/30 9:58 AM] Dan Seefeldt
>     Yes.
> ​[10/30 9:59 AM] Dan Seefeldt
>     The difference between production and tarball is that production build has it defined as 4.7.0 while the tarball build is 5.0.0.

> [10/30 10:40 AM] Davis Goodin
>   Got it, looks like TargetFrameworks is part of the key for NuGet to mash them together and fail: https://github.com/dagood/repro/tree/repro-static-restore-ref-mashup
> 
> Oh huh. Static restore is not necessary to repro.
>
> (I was wondering if we could also fix it by disabling static restore.)
>
> After mulling over what Tom's suggesting (and Wes suggested in the past) I think the real long term fix is to stop building all TFMs except net5.0... (so that we don't have this PackageReference to begin with) [referring to #1763]

> dseefeld: BTW, forcing System.Security.Cryptography.Cng to 4.7.0 fixed the issue.

4.7.0 isn't a problem because we should be able to build it in SBRP, so we should be able to leave it at that.

This info probably should have made it to the patch, or an issue linked to from the patch. (But we're rushing.)

Reply:
@dagood since you have a small reproducer, it seems worth filing a bug in https://github.com/NuGet/Home (even if it won't be fixed for our 5.0).

Reply:
Agreed, but I don't feel I have time to make it make enough sense for discussion. Filed https://github.com/dotnet/source-build/issues/1888 to track it though.

Reply:
FYI, I've filed the issue for this: https://github.com/NuGet/Home/issues/10368.